### PR TITLE
Fix 4176 reveal crash

### LINKED ISSUE #4176: Dafny crashing when revealing a non-static class method from outside the class
### Dafny version

4.1.0

### Code to produce this issue

```dafny
// class_test.dfy :

method test() {
    var c := new Class();
    reveal Class.P();
    assert c.P();
}

class Class {
  opaque function P() : bool { true }

  constructor () { }
}
```

### Command to run and resulting output

```code
dafny verify class_test.dfy
```

### What happened?

Upon running the above, Dafny is crashing with the following output:
```
% dafny verify ./test/class_test.dfy
/Users/ekanshdg/test/class_test.dfy(3,17): Error: accessing member 'reveal_P' requires an instance expression
/Users/ekanshdg/test/class_test.dfy(3,11): Error: call to instance method requires an instance
Unhandled exception. System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.)
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Dafny.CallStmt.get_NonSpecificationSubExpressions()+MoveNext() in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/AST/Statements/CallStmt.cs:line 73
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.Dafny.ProgramTraverser.Traverse(Statement stmt, String field, Object parent) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 923
   at Microsoft.Dafny.ProgramTraverser.<>c__DisplayClass17_0.<Traverse>b__2(Statement subStmt) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 925
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.Dafny.ProgramTraverser.Traverse(Statement stmt, String field, Object parent) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 923
   at Microsoft.Dafny.ProgramTraverser.<>c__DisplayClass17_0.<Traverse>b__2(Statement subStmt) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 925
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.Dafny.ProgramTraverser.Traverse(Statement stmt, String field, Object parent) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 923
   at Microsoft.Dafny.ProgramTraverser.Traverse(MemberDecl memberDeclaration, String field, Object parent) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 905
   at Microsoft.Dafny.ProgramTraverser.<>c__DisplayClass15_0.<Traverse>b__0(MemberDecl memberDecl) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 779
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.Dafny.ProgramTraverser.Traverse(TopLevelDecl topd) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 779
   at Microsoft.Dafny.ProgramTraverser.Traverse(List`1 topLevelDecls) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Generic/Util.cs:line 720
   at Microsoft.Dafny.Resolver.ResolveTopLevelDecls_Core(List`1 declarations, Graph`1 datatypeDependencies, Graph`1 codatatypeDependencies, String moduleName, Boolean isAnExport) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Resolver/Resolver.cs:line 2928
   at Microsoft.Dafny.Resolver.ResolveModuleDefinition(ModuleDefinition m, ModuleSignature sig, Boolean isAnExport) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Resolver/Resolver.cs:line 833
   at Microsoft.Dafny.Resolver.ResolveProgram(Program prog) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/Resolver/Resolver.cs:line 534
   at Microsoft.Dafny.Main.Resolve(Program program, ErrorReporter reporter) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/DafnyMain.cs:line 266
   at Microsoft.Dafny.Main.ParseCheck(IList`1 files, String programName, ErrorReporter reporter, Program& program) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyCore/DafnyMain.cs:line 216
   at Microsoft.Dafny.DafnyDriver.ProcessFilesAsync(IList`1 dafnyFiles, ReadOnlyCollection`1 otherFileNames, ErrorReporter reporter, Boolean lookForSnapshots, String programId) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyDriver/DafnyDriver.cs:line 403
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Microsoft.Dafny.DafnyDriver.ThreadMain(String[] args) in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyDriver/DafnyDriver.cs:line 124
   at Microsoft.Dafny.DafnyDriver.<>c__DisplayClass10_0.<Main>b__0() in /private/tmp/dafny-20230511-13185-uupd2y/dafny-4.1.0/Source/DafnyDriver/DafnyDriver.cs:line 102
   at System.Threading.Thread.StartCallback()
   ```

### What type of operating system are you experiencing the problem on?

Mac

### PR DESCRIPTION
**Pull request by EkanshdeepGupta**
Fixes #4176 by making generated reveal lemmas be static.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/IntegrationTests/IntegrationTests.csproj [hunk 1]
```diff
@@ -47,6 +47,16 @@
         <Content Include="..\..\node_modules\**\*.*" LinkBase="TestFiles\LitTests\LitTest\node_modules">
             <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
         </Content>
+        <Content Update="..\..\Test\git-issues\github-issue-4176.dfy">
```
Comment (MikaelMayer) [inline Source/IntegrationTests/IntegrationTests.csproj:50]: Please remove this change. We don't need this here, tests are discovered automatically. It was probably added by Rider.

File: Source/DafnyCore/Rewriters/OpaqueMemberRewriter.cs [hunk 2]
```diff
@@ -140,11 +140,11 @@ public OpaqueMemberRewriter(ErrorReporter reporter)
     }
     Method reveal;
     if (m is TwoStateFunction) {
-      reveal = new TwoStateLemma(m.RangeToken, m.NameNode.Prepend("reveal_"), m.HasStaticKeyword, new List<TypeParameter>(), new List<Formal>(), new List<Formal>(), new List<AttributedExpression>(),
+      reveal = new TwoStateLemma(m.RangeToken, m.NameNode.Prepend("reveal_"), true, new List<TypeParameter>(), new List<Formal>(), new List<Formal>(), new List<AttributedExpression>(),
```
Comment (MikaelMayer) [inline Source/DafnyCore/Rewriters/OpaqueMemberRewriter.cs:143]: Please also write a test to cover twostate opaque predicates too.

**Review by MikaelMayer — CHANGES_REQUESTED** [review]
Two changes to bring, but thank you very much for this contribution!

File: Source/DafnyCore/Rewriters/OpaqueMemberRewriter.cs (hunk 2 above)
Comment (EkanshdeepGupta) [reply Source/DafnyCore/Rewriters/OpaqueMemberRewriter.cs:143]: Done.

**Review by MikaelMayer — APPROVED** [review]
This is a great fix that surely @atomb  will appreciate too ! Looking forward to the test suite passing.

**Comment by MikaelMayer** [comment]
Still two more tests failing! But you are making great progress!

**Comment by atomb** [comment]
Oh, this is great! Requiring `reveal` to have an appropriate object in scope has been a significant difficulty for me in the past. And I'm glad it was so simple to fix!

**Review by MikaelMayer — APPROVED** [review]
Let's ship it! Great job.

**Pull request merged**